pkg/apis/kubernetes/v1/common.go

```diff
@@ -0,0 +1,56 @@
+package v1
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"go.anx.io/go-anxcloud/pkg/api"
+	"go.anx.io/go-anxcloud/pkg/api/types"
+	"go.anx.io/go-anxcloud/pkg/apis/internal"
+)
+
+const awaitCompletionPollInterval = 30 * time.Second
+
+var (
+	// ErrStateError is returned if a resource could not be provisioned (state "Error")
+	ErrStateError = errors.New("resource has state \"Error\"")
+
+	// ErrStateUnknown is returned if a resource has an unknown state
+	ErrStateUnknown = errors.New("resource has an unknown state")
+)
+
+type objectWithState interface {
+	types.Object
+	getState() internal.State
+}
+
+func awaitCompletion(ctx context.Context, a api.API, o objectWithState) error {
+	ticker := time.NewTicker(awaitCompletionPollInterval)
+	defer ticker.Stop()
+
+	for {
+		if err := a.Get(ctx, o); err != nil {
+			return fmt.Errorf("failed to get cluster: %w", err)
+		}
+
+		switch o.getState().ID {
+		case "0": // OK -> done
+			return nil
+		case "1": // Error
+			return ErrStateError
+		case "2": // Pending -> wait and try again
+			break
+		default:
+			return ErrStateUnknown
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+			continue
+		}
+	}
+}
```

Comment:
I like this, but maybe we could allow users to call this directly with any Object that implements `(lbaasv1).StateRetriever`

Reply:
I don't think that this is possible, as the function/method would need access to the Object which is not accessible from the embedded type, right?